src/rust/engine/serverset/src/lib.rs

```diff
@@ -218,67 +285,68 @@ impl<T: Clone + Send + Sync + 'static> Serverset<T> {
   /// first server to become healthy after all are unhealthy).
   ///
   pub fn next(&self) -> BoxFuture<(T, HealthReportToken), String> {
-    let now = Instant::now();
     let mut inner = self.inner.lock();
-    let server_count = inner.connections.len();
-
-    let mut earliest_future = None;
-    for _ in 0..server_count {
-      let i = inner.next % server_count;
-      inner.next = inner.next.wrapping_add(1);
-      let server = &inner.connections[i];
-      let unhealthy_info = &server.unhealthy_info;
-      if let Some(ref unhealthy_info) = unhealthy_info {
-        // Server is unhealthy. Note when it will become healthy.
-
-        let healthy_at = unhealthy_info.healthy_at();
-
-        if healthy_at > now {
-          let healthy_sooner_than_previous = if let Some((_, previous_healthy_at)) = earliest_future
-          {
-            previous_healthy_at > healthy_at
-          } else {
-            true
-          };
-
-          if healthy_sooner_than_previous {
-            earliest_future = Some((i, healthy_at));
-          }
-          continue;
-        }
+
+    if inner.can_make_more_connections() {
+      // Find a healthy server without a connection, connect to it.
+      if let Some(ret) = inner.connect() {
+        inner.next_connection = inner.next_connection.wrapping_add(1);
```

Comment:
It's not clear why we increment `next_connection` here... the connection will be added at the end of the list, but we don't know where the index is pointing. If this is necessary would add a comment, otherwise, maybe remove to keep things clear?

Reply:
Done